Mathlib/NumberTheory/NumberField/ProductFormula.lean

```diff
@@ -0,0 +1,111 @@
+/-
+Copyright (c) 2024 Fabrizio Barroero. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Fabrizio Barroero
+-/
+import Mathlib.NumberTheory.NumberField.FinitePlaces
+
+/-!
+# The Product Formula for number fields
+In this file we prove the Product Formula for number fields: for any non-zero element `x` of a
+number field `K`, we have `∏|x|ᵥ=1` where the product runs over the equivalence classes of absoulte
+values of `K` and the `|⬝|ᵥ` are suitably normalized.
+
+## Main Results
+* `NumberField.FinitePlace.prod_eq_inv_abs_norm`: for any non-zero element `x` of a number field
+`K`, the product `∏|x|ᵥ` of the absolute values of `x` associated to the finite places of `K` is
+equal to the inverse of the norm of `x`.
+* `NumberField.prod_abs_eq_one`: for any non-zero element `x` of a number field `K`, we have
+`∏|x|ᵥ=1` where the product runs over the equivalence classes of absoulte values of `K`.
+
+## Tags
+number field, embeddings, places, infinite places, finite places, product formula
+-/
+
+namespace NumberField
+
+variable {K : Type*} [Field K] [NumberField K]
+
+open Algebra Classical FinitePlace Function Ideal IsDedekindDomain
+  IsDedekindDomain.HeightOneSpectrum
+
+/-- For any non-zero `x` in `𝓞 K`, the prduct of `w x`, where `w` runs over `FinitePlace K`, is
+equal to the inverse of the absolute value of `Algebra.norm ℤ x`. -/
+theorem FinitePlace.prod_eq_inv_abs_norm_int {x : 𝓞 K} (h_x_nezero : x ≠ 0) :
+    ∏ᶠ w : FinitePlace K, w x = (|norm ℤ x| : ℝ)⁻¹ := by
+  convert_to ∏ᶠ v : HeightOneSpectrum (𝓞 K), ‖embedding v x‖ = |↑(norm ℤ x)|⁻¹
+  · exact (finprod_eq_of_bijective maximalIdeal ((bijective_iff_existsUnique _).mpr
+      <| fun v ↦ ⟨mk v, maximalIdeal_mk v, fun _ a ↦ by rw [← a, mk_maximalIdeal]⟩)
```

Comment:
`bijective_iff_existsUnique` is a very roundabout way to go about this. You have `maximalIdeal_mk` and `mk_maximalIdeal` which are exactly the ingredients to define
```lean
noncomputable def NumberField.FinitePlace.equivHeightOneSpectrum :
    FinitePlace K ≃ HeightOneSpectrum (𝓞 K) where
  toFun := maximalIdeal
  invFun := mk
  left_inv := mk_maximalIdeal
  right_inv := maximalIdeal_mk
```

(probably this should be in `NumberTheory.NumberField.FinitePlaces`) and then lemmas
like `maximalIdeal_injective` and `maximalIdeal_inj` aren't even really necessary because they're just `(equivHeightOneSpectrum K).injective` and `(equivHeightOneSpectrum K).injective.eq_iff`. Similarly this proof becomes
```lean
  · exact (finprod_eq_of_bijective maximalIdeal (equivHeightOneSpectrum K).bijective
      (fun w ↦ (norm_embedding_eq w (x : K)).symm))
```
We have tons of API for `equiv` so you should use it.